client/authority-discovery/src/worker.rs

```diff
@@ -495,51 +482,73 @@ where
 		let remote_addresses: Vec<Multiaddr> = values
 			.into_iter()
 			.map(|(_k, v)| {
-				let schema::SignedAuthorityAddresses { signature, addresses } =
-					schema::SignedAuthorityAddresses::decode(v.as_slice())
+				let schema::SignedAuthorityRecord { record, auth_signature, peer_signature } =
+					schema::SignedAuthorityRecord::decode(v.as_slice())
 						.map_err(Error::DecodingProto)?;
 
-				let signature = AuthoritySignature::decode(&mut &signature[..])
+				let auth_signature = AuthoritySignature::decode(&mut &auth_signature[..])
 					.map_err(Error::EncodingDecodingScale)?;
 
-				if !AuthorityPair::verify(&signature, &addresses, &authority_id) {
+				if !AuthorityPair::verify(&auth_signature, &record, &authority_id) {
 					return Err(Error::VerifyingDhtPayload)
 				}
 
-				let addresses = schema::AuthorityAddresses::decode(addresses.as_slice())
-					.map(|a| a.addresses)
-					.map_err(Error::DecodingProto)?
-					.into_iter()
-					.map(|a| a.try_into())
-					.collect::<std::result::Result<_, _>>()
-					.map_err(Error::ParsingMultiaddress)?;
-
+				let mut addresses: Vec<Multiaddr> =
+					schema::AuthorityRecord::decode(record.as_slice())
+						.map(|a| a.addresses)
+						.map_err(Error::DecodingProto)?
+						.into_iter()
+						.map(|a| a.try_into())
+						.collect::<std::result::Result<_, _>>()
+						.map_err(Error::ParsingMultiaddress)?;
+
+				let get_peer_id = |a: &Multiaddr| match a.iter().last() {
+					Some(multiaddr::Protocol::P2p(key)) => PeerId::from_multihash(key).ok(),
+					_ => None,
+				};
+
+				// Ignore [`Multiaddr`]s without [`PeerId`] or with own addresses.
+				let addresses: Vec<Multiaddr> = addresses
+					.drain(..)
+					.filter_map(|a| {
+						get_peer_id(&a)
+							.and_then(|p| if p != local_peer_id { Some(a) } else { None })
+					})
+					.collect();
+
+				let remote_peer_id = single(addresses.iter().map(get_peer_id))
```

Comment:
Don't we doing the `get_peer_id` already above?

Reply:
Nope.